// reactivated/build.renderer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var esbuild_1 = __importDefault(require("@linaria/esbuild"));
var esbuild_plugin_1 = require("@vanilla-extract/esbuild-plugin");
var esbuild = __importStar(require("esbuild"));
var esbuild_plugin_import_glob_1 = __importDefault(require("esbuild-plugin-import-glob"));
var server = null;
var CACHE_KEY = "".concat(process.cwd(), "/node_modules/_reactivated/renderer.js");
var production = process.env.NODE_ENV === "production";
var identifiers = production ? "short" : "debug";
esbuild
    .build({
    stdin: {
        contents: "\n                export {server} from \"reactivated/renderer\";\n            ",
        resolveDir: process.cwd(),
        loader: "ts",
    },
    minify: production,
    bundle: true,
    platform: "node",
    outfile: "./node_modules/_reactivated/renderer.js",
    sourcemap: true,
    target: "es2018",
    preserveSymlinks: true,
    watch: production === true
        ? false
        : {
            onRebuild: function () {
                restartServer();
            },
        },
    plugins: [
        (0, esbuild_plugin_import_glob_1.default)(),
        // We manually pass in identifiers because the client is not
        // minified by esbuild but the renderer is, so class names could
        // differ.
        // Instead of set it manually instead of relying on minification
        // settings.
        (0, esbuild_plugin_1.vanillaExtractPlugin)({ identifiers: identifiers }),
        (0, esbuild_1.default)({ sourceMap: true }),
    ],
})
    .then(function () {
    if (production === false) {
        restartServer();
    }
})
    .catch(function () { return process.exit(1); });
var restartServer = function () {
    if (server != null) {
        server.close();
    }
    delete require.cache[CACHE_KEY];
    server = require(CACHE_KEY).server;
};
