// reactivated/build.client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var esbuild_1 = __importDefault(require("@linaria/esbuild"));
var esbuild_plugin_1 = require("@vanilla-extract/esbuild-plugin");
var esbuild = __importStar(require("esbuild"));
var esbuild_plugin_import_glob_1 = __importDefault(require("esbuild-plugin-import-glob"));
var entryNames = process.argv.slice(2);
var entryPoints = Object.fromEntries(entryNames.map(function (entry) { return [entry, "./client/".concat(entry, ".tsx")]; }));
var production = process.env.NODE_ENV === "production";
var identifiers = production ? "short" : "debug";
var env = {
    NODE_ENV: production ? "production" : "development",
    BUILD_VERSION: process.env.BUILD_VERSION,
    TAG_VERSION: process.env.TAG_VERSION,
};
esbuild
    .build({
    entryPoints: entryPoints,
    bundle: true,
    // We use terser to minify because esbuild breaks safari sourcemaps.
    // It's likely a Safari bug, but terser seems to work for some reason.
    minify: false,
    // Related to sourcemaps as well in Safari.
    legalComments: "none",
    platform: "browser",
    outdir: "./static/dist",
    sourcemap: true,
    target: "es2018",
    preserveSymlinks: true,
    watch: production === false,
    external: ["moment", "@client/generated/images"],
    define: {
        // You need both. The one from the stringified JSON is not picked
        // up during the build process.
        "process.env.NODE_ENV": production ? '"production"' : '"development"',
        process: JSON.stringify({ env: env }),
        // Redux persist needs this.
        global: "{}",
    },
    plugins: [
        (0, esbuild_plugin_import_glob_1.default)(),
        // We manually pass in identifiers because the client is not
        // minified by esbuild but the renderer is, so class names could
        // differ.
        // Instead of set it manually instead of relying on minification
        // settings.
        (0, esbuild_plugin_1.vanillaExtractPlugin)({ identifiers: identifiers }),
        (0, esbuild_1.default)({ sourceMap: true, esbuildOptions: { sourcemap: "inline" } }),
    ],
})
    .catch(function () { return process.exit(1); });
